### DISCUSSION TITLE
[@types/react] Context Provider typing forbids to use it as a JSX element type since v18.2.8

### DISCUSSION THREAD

**Discussion by birdofpreyru** [DefinitelyTyped/DefinitelyTyped#67069] [Issues with a @types package]
Hey, I am migrating to TypeScript (v5.2.2) a project that uses React (v18.2.0) Context API. Nothing crazy, I just do the standard context use:
```ts
const Context = createContext<ValueType>(initialValue);

function myFunc() {
  // Some calculation of `myValue` constant omitted.
  return <Context.Provider value={myValue}>{children}</Context.Provider>;
}
```
with `@types/react@18.2.7` it is fine; but starting with `@types/react@18.2.8`, and up to the latest v18.2.22 it does not compile with the error:
> 'Context.Provider' cannot be used as a JSX component.
>  Its return type 'ReactNode' is not a valid JSX element.
>    Type 'undefined' is not assignable to type 'Element | null'.

If I just remove `undefined` from `type ReactNode` definition, then it compiles fine; and then looking to the definitions it seems to work with v18.2.7 because there `ExoticComponent` interface (which is extended by context provider) has `(props: P): (ReactElement|null);` and later it was changed into `(props: P): ReactNode;`.

Thus, I believe, it is a bug in definitions, and who owns them should look into fixing it.

[a comment by an automated account is not shown]

**Comment by birdofpreyru** [comment]
Another problem I have with `@types/react` v18.2.6 and above is this:
```ts
export function Test(Component: React.ComponentType): JSX.Element {
  return <Component />;
}
```
:point_up: tells me
> _Type 'Element' is missing the following properties from type 'ReactElement<any, any>': type, props, key_

works fine with `@types/react` v18.2.5. For some reason with v18.2.6 and up it thinks its type is `interface Element` from `typescript/lib/lib.dom.d.ts`, rather than `JSX.Element` :(

**Comment by eps1lon** [comment]
Make sure you only have a single version of `@types/react` installed following: https://solverfox.dev/writing/duplicate-types-react

If the issue persists, please provide a minimal, cloneable reproduction.

**Reply by birdofpreyru** [reply]
Here is the repro &mdash; [repro.zip](https://github.com/DefinitelyTyped/DefinitelyTyped/files/12920576/repro.zip)
Just unzip and do `npm install && npm run build`.
Somehow, the issue depends on `compilerOptions.module` setting in the TypeScript config. With `ES6`, and other settings it is error, with `CommonJS` it is not.

**Comment by deniz-blue** [comment]
I just had this issue with React 19.1.0 and types 19.1.2 - changing `moduleResolution` to `bundler` fixed it
Don't know if its the same issue but hopefully this info will help anyone else in the future

**Comment by karanuday** [comment]
I've been struggling with this issue for an entire day now. Absolutely no luck with suggested/recommended approaches.

**Approaches tried:**
- Changing `moduleResolution` to bundler isnt an option in my case since it breaks some internal deps in my monorepo. 
- As suggested in multiple other reproductions of this issue, also made sure I only have a single version for `types/react` and `types/react-dom`.

The only approach that worked was to upgrade React to. v19 but keep types at v18 which is not something I'd like to do as this could cause incompatibility issues down the line. 

If anybody has an approach that works for a NextJS App with React v19, would be highly appreciated!

**Reply by birdofpreyru** [reply]
Hey, I've long forgot about this issue I posted, I don't remember how I solved it, but it apparently is not an issue for me anymore with the latest React and its type definitions. Though, I have in TypeScript `compilerOptions` of all my React projects:
```ts
"compilerOptions": {
  "module": "ESNext",
  "moduleResolution": "Node"
}
```
perhaps this is the magical combination with which it works :)

**Comment by swiing** [comment]
I also had similar issue. Bumping `@types/react` to latest (v19.1.12 as of time of writing) solved it.

**Comment by preco21** [comment]
For anyone facing this issue in combination with [Emotion](https://emotion.sh/):

Refer to https://github.com/emotion-js/emotion/issues/3049 for details.

In my case, the issue persisted even after upgrading `@types/react` to the latest version and was resolved by upgrading `@emotion/react` to the latest version. (At least, version: `@emotion/react@11.11.1`)
